### ISSUE TITLE
ARM64 (aarch64) on Windows 10: sh.exe "not compatible with the version of windows you're running"

### ISSUE THREAD

**Issue by MikeFarrington** [git-for-windows/git#5292]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
git version 2.47.1.windows.1
cpu: aarch64
built from commit: 2cd22437f64229935dc564db969cbcbfed5e9045
sizeof-long: 4
sizeof-size_t: 8
shell-path: D:/git-sdk-arm64-build-installers/usr/bin/sh          <------ this is weird, not a drive or folder on my system
feature: fsmonitor--daemon
libcurl: 8.11.0
OpenSSL: OpenSSL 3.2.3 3 Sep 2024
zlib: 1.3.1
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
Windows 10 (22H2) on 64-bit ARMv7 (Qualcomm Snapdragon 835 CPU)

Microsoft Windows [Version 10.0.19045.5198]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

Mostly defaults.  VSCode for GUI, REBASE by default

```
Editor Option: VisualStudioCode
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: ConHost
Git Pull Behavior Option: Rebase
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable FSMonitor: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

I am running an older ARM on Windows CPU which isn't getting official Windows 11 support from Microsoft but still works well with just about every natively compiled application for Windows on ARM.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

I've tried both CMD and PowerShell.  Bash will not launch due to error being reported.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
.\sh.exe
git fetch
```
 - What did you expect to occur after running these commands?

```
Git to work as expected, without throwing errors about the version of Windows
```

 - What actually happened instead?

```
PS C:\Program Files\Git\bin> .\sh.exe
error launching git: This version of %1 is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher.
```
```
PS C:\MyGitProjectFolder> git fetch
error: cannot spawn C:/Program Files/Git/usr/bin/sh.exe: Function not implemented
error: cannot spawn bash: Function not implemented
fatal: could not read Username for 'https://github.com': No such file or directory
```

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

```
Not related to repository
```

**Comment by dscho** [comment]
This is sadly expected. Windows 10 only runs i686 code (also known as x86, i.e. 32-bit AMD/Intel) in emulation, and not x86_64 (also known as x64, i.e. 64-bit AMD/Intel). However, for technical reasons Git Bash does not yet support Windows/ARM64 natively.

I tried to clarify this in [this comment](https://github.com/git-for-windows/git/discussions/5276#discussioncomment-11386319).

**Comment by MikeFarrington** [comment]
> This is sadly expected. Windows 10 only runs i686 code (also known as x86, i.e. 32-bit AMD/Intel) in emulation, and not x86_64 (also known as x64, i.e. 64-bit AMD/Intel). However, for technical reasons Git Bash does not yet support Windows/ARM64 natively.
> 
> I tried to clarify this in [this comment](https://github.com/git-for-windows/git/discussions/5276#discussioncomment-11386319).

Ah, that explains it.  My OS/CPU only supports the WOW emulation of 32-bit x86 code, not amd64 code.  I wonder if the x86/32-bit bash shell should be included for the very few like me who still have these ARMv7 WOA computers in the wild?

**Comment by dscho** [comment]
While I am sympathetic, I can indeed not justify working on it. But I can give you pointers to get started:

- you could start by extracting Portable-Git-32-bit,
  - then removing its `mingw32` folder and replacing it with the `clangarm64` folder of Portable-Git-arm64,
  - then starting Git Bash
  - then adjust `PATH` by replacing `mingw32` with `clangarm64` in it.

A "full" solution would be much more involved, and would assume that you can use `git clone` from Portable-Git-arm64. The strategy would then be to:
- fork `git-sdk-32`,
- clone it,
- start a Git SDK Bash by launching the `git-bash.exe` contained in the top-level directory of that worktree
- remove all `mingw-w64-i686-*` packages from it via `pacman -R <package>...`,
- add the `clangarm64` Pacman repository by editing `/etc/pacman.conf`,
- install the `mingw-w64-clang-aarch64-* versions of the just-removed packages,
- re-run `git-bash.exe` [so that the `PATH` includes `/clangarm64/bin/`](https://github.com/git-for-windows/MINGW-packages/blob/4f83a265e56b2a12010d00829fbc71646546c853/mingw-w64-git/git-wrapper.c#L139-L152),
- publish the fork,
- work out all the kinks 😉

**Comment by dscho** [comment]
> Unfortunately, the x86 version isn't working under emulation. It just sort of hangs and does nothing on a 'git fetch'.

This is a bit worrisome, as the same root cause may potentially prevent a lot of Git Bash's stuff to work, anyway, in which case all of the effort to get it to run would be in vain.

**Comment by jeremyd2019** [comment]
I was thinking about this too, but figured it was a small enough userbase to not be worth it.  Did you backport the hang fix to your i686 msys2-runtime-3.3 fork here?

Maybe you could add a check to block install on Windows 10 to the ARM64 installer if you don't want to support it?

**Comment by dscho** [comment]
> Did you backport the hang fix to your i686 msys2-runtime-3.3 fork here?

No... But I'd gladly accept a PR (@MikeFarrington hint, hint😀 )

> Maybe you could add a check to block install on Windows 10 to the ARM64 installer if you don't want to support it?

Sure, but I do not want to invest energy in preventing something that highly interested & motivated users may still bring about.

**Comment by jeremyd2019** [comment]
> > Maybe you could add a check to block install on Windows 10 to the ARM64 installer if you don't want to support it?
> 
> Sure, but I do not want to invest energy in preventing something that highly interested & motivated users may still bring about.

It seems to me that the small percentage of ARM64 users who need i686 MSYS2 wouldn't be worth switching ARM64 entirely over to i686 (with its issues including lack of upstream support and address space conflicts), so if somebody did make an i686+ARM64 version, it would probably be a different installer than the x86_64+ARM64 one (which would still not work properly on Windows 10).

**Comment by jeremyd2019** [comment]
> > Did you backport the hang fix to your i686 msys2-runtime-3.3 fork here?
> 
> No... But I'd gladly accept a PR (@MikeFarrington hint, hint😀 )

If you don't know what I'm talking about, it's https://github.com/msys2/msys2-runtime/compare/bb6258c..fbc1019 (but it looks like [1f3be1e on](https://github.com/msys2/msys2-runtime/compare/3cba82a..msys2-3_3_6-release) are not in [git-for-windows/msys2-runtime's 3.3 branch](https://github.com/git-for-windows/msys2-runtime/commits/msys2-gfw-3.3-branch/))

**Comment by jeremyd2019** [comment]
This seems to work for me:
1. I already had [Git-2.47.1-32-bit.exe](https://github.com/git-for-windows/git/releases/download/v2.47.1.windows.1/Git-2.47.1-32-bit.exe) installed
2. I downloaded [Git-2.47.1-arm64.tar.bz2](https://github.com/git-for-windows/git/releases/download/v2.47.1.windows.1/Git-2.47.1-arm64.tar.bz2)
3. In git bash, I did `tar -C / -jxvf Git-2.47.1-arm64.tar.bz2 clangarm64/`
4. I exited and restarted git bash
5. git --version --build-options shows I'm using the arm64 git, and git fetch/stash/pull/status all seem to work (including dealing with a conflict on `git stash pop`)

**Comment by jeremyd2019** [comment]
Actually, I just noticed something: for some reason, the system `gitconfig` file lives in `/etc` on x86 installs, but it seems the clangarm64 git expects it in `/clangarm64/etc` ?  In any event, `git config --list` is empty after that procedure.

**Comment by dscho** [comment]
Okay, this isn't anything we can support realistically. 32-bit support has been dropped except for MinGit, and a fully native ARM64 version of the MSYS2 runtime is also still quite a bit in the future. Once it arrives, though, Windows/ARM64 10 _should_ work without problems.
